src/NHibernate/Type/GenericBagType.cs

```diff
@@ -70,5 +72,12 @@ public override object Instantiate(int anticipatedSize)
 		{
 			return anticipatedSize <= 0 ? new List<T>() : new List<T>(anticipatedSize + 1);
 		}
+
+		protected override bool AreCollectionElementsEqual(IEnumerable original, IEnumerable target)
+		{
+			var first = (IEnumerable<T>)original;
+			var second = (IEnumerable<T>) target;
+			return first.All(element => first.Count(x => x.Equals(element)) == second.Count(x => x.Equals(element)));
```

Comment:
Testing general count equality is required, otherwise a bad `true` will be yielded if `target` is "wider" than `original` by a few elements not present in `original`. So it should be:

```C#
return first.Count() == second.Count() &&
    first.All(element => first.Count(x => x.Equals(element)) == second.Count(x => x.Equals(element)));
```